### ISSUE TITLE
[BUG] Treegrid is loading an eternity for huge amounts of data

### ISSUE THREAD

**Issue by luwol03** [inventree/InvenTree#3438]
**Problem**
If you have deeply nested storage locations and viewing them via the tree view has two disadvantages. 
1. It needs a very long time to load
2. I'ts not very readable.
![image](https://user-images.githubusercontent.com/60048565/181996528-df8e070f-609e-4f69-98c3-8d7ce7a95bf9.png)
(These naming makes no sense of an actual naming, but this was just the result of some demo data from me. In my real storage its not nested like that, its more like Room/rack/shelve/box/box location (one box is devided into front/middle/back). So my tree is nested very deeply and each branch contains a lot of items. And imagine a storage with 4 rooms each 12 racks each 32 boxes each splitted into 3 sections)

**Suggested solution**
Add a max tree depth and then lazy load the items if we try to open that specific StorageLocation.

**Describe alternatives you've considered**
-

**Examples of other systems**
-

**Do you want to develop this?**
If this is really wanted I would need some hints where to start.

**Comment by SchrodingersGat** [comment]
@luwol03 how long is the API request taking to load all this data?

**Comment by luwol03** [comment]
It's taking about 10 minutes for me. I generated the data with this script:
```py
def generate_nested(parent, loc):
    if len(loc) == 0:
        return

    cnt, name = loc.pop(0)
    print(parent.pathstring)
    for i in range(cnt):
        s = StockLocation.create(api, data={"parent": parent.pk, "name": f"{name} {i}"})
        generate_nested(s, loc.copy())


data = [
    (4, "Room"),
    (12, "Rack"),
    (32, "Box"),
    (3, "Section")
]

generate_nested(StockLocation(api, 305), data)
```

### Timing:
Interesting, there are several requests with different timings.
- `GET /api/stock/location/tree/` - `0,294s`
- `GET /api/stock/location/?parent=305&cascade=true` - `23.608s`

Also interesting is, that the page needs about 10 minutes to load the tree. So there must be also some client side overload.

**Comment by SchrodingersGat** [comment]
10 minutes is an eternity. How many locations in that dataset?

**Comment by luwol03** [comment]
You can calculate it via 4*12*32*3=4608.

**Comment by SchrodingersGat** [comment]
> Add a max tree depth

Do you mean that items below a certain "depth" in the tree would not be displayed in the breadcrumb tree?

**Comment by luwol03** [comment]
> Do you mean that items below a certain "depth" in the tree would not be displayed in the breadcrumb tree?

Yes, and if I click on the triangle infront of the name it will lazy load the next level of the tree.

**Comment by SchrodingersGat** [comment]
This function is most likely the culprit for the extreme loading times:

https://github.com/inventree/InvenTree/blob/d4eae9da116be3a1770ee757db0acdd25b2d471c/InvenTree/templates/js/dynamic/nav.js#L152

While the actual database query is pretty quick (~300ms, though this could probably be improved!) - this javascript function has some nasty nested loops and is in need of a major refactor.

**Comment by luwol03** [comment]
It looks like the the problem is not in this function. It only takes about 50ms (measured with `console.time`). I will try to find a solution for this problem.

But what do you think of making everything deeper than 2/3 levels collapsed and not open by default?

**Comment by SchrodingersGat** [comment]
> But what do you think of making everything deeper than 2/3 levels collapsed and not open by default?

I'd like to see an implementation of this, not sure exactly how it would go but am interested!

**Comment by luwol03** [comment]
@SchrodingersGat I would be happy if you could point me in the right direction about the jquery treeview function. What is that function calling? I cannot find anything useful.

**Comment by luwol03** [comment]
@SchrodingersGat I found out, that Inventree uses [bootstrap-table](https://bootstrap-table.com) for the tree view in combination with the [tree grid](https://examples.bootstrap-table.com/#extensions/treegrid.html) plugin, right?
 
I did a little profiling the first 30s and It looks like this plugin is responsible for these long waiting times. Here are my results:
![image](https://user-images.githubusercontent.com/60048565/182046965-9473bc65-cd1a-49e7-87f4-3297898c954d.png)
If you want to take a look yourself, here is my [1_Profile-20220731T233104.json.zip](https://github.com/inventree/InvenTree/files/9229792/1_Profile-20220731T233104.json.zip). You can import it into a chrome browser and inspect the long running functions.

I think this looks like an upstream issue.

**Comment by SchrodingersGat** [comment]
Ah sorry when you said "breadcrumb tree" I thought you meant the navigation tree at the top of the page:

![image](https://user-images.githubusercontent.com/10080325/182062585-a6e6cf2c-ac7e-4b64-8d9e-4ecb486fb252.png)

I see that it is the part category table which is rendering slowly. I can definitely replicate that when rendering the table in "tree view" mode.

Lazy-loading the sub items would be a good idea. We do something like this with the BOM table:

![image](https://user-images.githubusercontent.com/10080325/182062726-4e10585e-db5d-4379-acf0-03a33baabb54.png)

You can have a look at the `loadBomTable` method in `bom.js` to see how this is performed.

In particular:

https://github.com/inventree/InvenTree/blob/099489ce792fec91bde08dcfdff4146a6e948fdf/InvenTree/templates/js/translated/bom.js#L1157

**Comment by SchrodingersGat** [comment]
> Why does Inventree use jQuery? Isn't that more like a script which bloats the load size as many tasks can be done nowadays with plain js

I'm not sure what answer you want here :p It uses jQuery because that's what I chose when I started developing it (and, to be fair, knew absolutely nothing about front-end dev). It has served us well, and is pretty minimal in terms of overhead.


> what is the state of switching to react as referenced here: https://github.com/inventree/InvenTree/issues/2789 ?

It is in the long term planning stage. There is a long way to go before we start such a major front-end overhaul, and it will be a *huge* undertaking that will probably block any other effort while it is being developed.

**Comment by luwol03** [comment]
Thank you for your answers. I totally understand that switching to react will be a major change. 

Sorry for my confusion, I thought the whole time, that I understand something not correctly because the function where I focused on is not responsible for what I was searching for. Yes you're right, not the breadcrumb. I mean the table treegrid.

With #3443 you reduced the request time (`GET /api/stock/location/?parent=1&cascade=true`) from 23s to 4s with the data generated as shown above with the script while viewing the outermost location

I tried to implement the lazy loading in #3451. Let me know what you think.
